shared/js/intl_helper.js

```diff
@@ -0,0 +1,170 @@
+(function(global) {
+  'use strict';
+
+  /*************************
+
+  IntlHelper.define('shortDate', 'datetime', {
+    year: 'numeric',
+    month: 'long',
+  });
+
+  function setValue() {
+    var formatter = IntlHelper.get('shortDate');
+    console.log(formatter.format(new Date()));
+  }
+
+  setValue();
+
+  IntlHelper.observe('shortDate', setValue);
+  
+   ****************************/
+  var objectCache;
+
+  const knownObjects = {
+    'datetime': {
+      set: function(options) {
+        const customOptions = Object.assign({}, options);
+        if (options.hour) {
+          customOptions.hour12 = navigator.mozHour12;
+        }
+        return Intl.DateTimeFormat(
+          navigator.languages,
+          customOptions
+        );
+      },
+      isAffected: function(obj, reason) {
+        if (reason !== 'timeformatchange') {
+          return true;
+        }
+
+        return 'hour' in obj.options;
+      }
+    },
+    'number': {
+      set: function(options) {
+        return Intl.NumberFormat(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+    'collator': {
+      set: function(options) {
+        return Intl.Collator(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+  };
+
+  const affectedTypes = {
+    'languagechange': ['datetime', 'number', 'collator'],
+    'timeformatchange': ['datetime'],
+  };
```

Comment:
Can we consider removing `affectedTypes`?  You can achieve the same thing via `isAffected` (which I really like because it encapsulates the logic inside of `knownObjects`) and the code could be simpler and easier to maintain.